From 2d34345e04a29db258baa955dfe2854cda2e0646 Mon Sep 17 00:00:00 2001
From: Pavankumar Nandeshwar <pnandesh@qti.qualcomm.com>
Date: Mon, 2 Jun 2025 17:50:30 +0530
Subject: [PATCH 03/21] UPSTREAM: DP: Remove forced alignment and use complete
 data size for map/unmap

Remove forced alignment and use complete data size for map/unmap
as the former is not neccessary.
Since, the allocated size of buffer is already DP_RX_BUFFER_SIZE,
which is the same length used for doing dma_map / dma_unmap
there is no need to read the same length again from skb.
This will avoid unnecessary dereferencing of skb in rx path.

Patch-dependency : None

Signed-off-by: Pavankumar Nandeshwar <pnandesh@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/dp.c          |  3 ++-
 drivers/net/wireless/ath/ath12k/dp_rx.c       | 17 ++++-------------
 drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c | 11 ++++-------
 3 files changed, 10 insertions(+), 21 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 21ce682..5c9774a 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1290,7 +1290,8 @@ static void ath12k_dp_cc_cleanup(struct ath12k_base *ab)
 				continue;
 
 			ath12k_core_dma_unmap_single(ab->dev, desc_info[j].paddr,
-						     skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
+						     DP_RX_BUFFER_SIZE,
+						     DMA_FROM_DEVICE);
 			dev_kfree_skb_any(skb);
 		}
 	}
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 47ae08c..22290f4 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -343,24 +343,15 @@ int ath12k_dp_rx_bufs_replenish(struct ath12k_dp *dp,
 
 	while (num_remain > 0) {
 #ifdef CPTCFG_MAC80211_SFE_SUPPORT
-		skb = netdev_alloc_skb_fast(NULL, DP_RX_BUFFER_SIZE +
-					    DP_RX_BUFFER_ALIGN_SIZE);
+		skb = netdev_alloc_skb_fast(NULL, DP_RX_BUFFER_SIZE);
 #else
-		skb = dev_alloc_skb(DP_RX_BUFFER_SIZE +
-				    DP_RX_BUFFER_ALIGN_SIZE);
+		skb = dev_alloc_skb(DP_RX_BUFFER_SIZE);
 #endif
 		if (!skb)
 			break;
 
-		if (!IS_ALIGNED((unsigned long)skb->data,
-				DP_RX_BUFFER_ALIGN_SIZE)) {
-			skb_pull(skb,
-				 PTR_ALIGN(skb->data, DP_RX_BUFFER_ALIGN_SIZE) -
-				 skb->data);
-		}
 #ifndef CONFIG_IO_COHERENCY
-		paddr = dma_map_single(dp->dev, skb->data,
-				       skb->len + skb_tailroom(skb),
+		paddr = dma_map_single(dp->dev, skb->data, DP_RX_BUFFER_SIZE,
 				       DMA_FROM_DEVICE);
 		if (dma_mapping_error(dp->dev, paddr))
 			goto fail_free_skb;
@@ -396,7 +387,7 @@ int ath12k_dp_rx_bufs_replenish(struct ath12k_dp *dp,
 	goto out;
 
 fail_dma_unmap:
-	ath12k_core_dma_unmap_single(dp->dev, paddr, skb->len + skb_tailroom(skb),
+	ath12k_core_dma_unmap_single(dp->dev, paddr, DP_RX_BUFFER_SIZE,
 				     DMA_FROM_DEVICE);
 fail_free_skb:
 	dev_kfree_skb_any(skb);
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
index 7030428..b6e6186 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
@@ -1447,7 +1447,7 @@ try_again:
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
 		ath12k_core_dma_unmap_single(partner_dp->dev, desc_info->paddr,
-					     msdu->len + skb_tailroom(msdu),
+					     DP_RX_BUFFER_SIZE,
 					     DMA_FROM_DEVICE);
 
 		num_buffs_reaped[device_id]++;
@@ -1820,7 +1820,7 @@ err_free_desc:
 	list_add_tail(&desc_info->list, &dp->rx_desc_free_list);
 	spin_unlock_bh(&dp->rx_desc_lock);
 err_unmap_dma:
-	ath12k_core_dma_unmap_single(ab->dev, buf_paddr, defrag_skb->len + skb_tailroom(defrag_skb),
+	ath12k_core_dma_unmap_single(ab->dev, buf_paddr, DP_RX_BUFFER_SIZE,
 				     DMA_TO_DEVICE);
 	return ret;
 }
@@ -2068,10 +2068,8 @@ ath12k_wifi7_dp_process_rx_err_buf(struct ath12k_pdev_dp *dp_pdev,
 
 	list_add_tail(&desc_info->list, used_list);
 
-	ath12k_core_dma_unmap_single(ab->dev, desc_info->paddr,
-				     msdu->len + skb_tailroom(msdu),
+	ath12k_core_dma_unmap_single(ab->dev, desc_info->paddr, DP_RX_BUFFER_SIZE,
 				     DMA_FROM_DEVICE);
-
 	if (drop) {
 		dev_kfree_skb_any(msdu);
 		return 0;
@@ -2658,8 +2656,7 @@ int ath12k_wifi7_dp_rx_process_wbm_err(struct ath12k_dp *dp,
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
 		ath12k_core_dma_unmap_single(partner_dp->dev, desc_info->paddr,
-					     msdu->len + skb_tailroom(msdu),
-					     DMA_FROM_DEVICE);
+					     DP_RX_BUFFER_SIZE, DMA_FROM_DEVICE);
 
 		num_buffs_reaped[device_id]++;
 		total_num_buffs_reaped++;
-- 
2.34.1

